designs/generic-git-provider-gitops.md

```diff
@@ -0,0 +1,205 @@
+```yaml
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: GitOpsConfig
+metadata:
+  name: my-generic-git-provider-gitops
+spec:
+  flux:
+    fluxSystemNamespace: ""
+    clusterConfigPath: ""
+    git:
+      repositoryUrl: myClusterGitopsRepo
+      username: myGitProviderUserName
+      passwordEnvironmentVariable: MY_PASSWORD_ENV_VAR
+      privateKeyFile: '/path/to/my/private_key.rsa'
+      branch: branchToUseInRemoteRepo
+```
+
+### Git Configuration Structs
+The `Flux` configuration will be modified to contain generic flux options `FluxSystemNamespace`, `Branch`, and `ClusterConfigPath`, which will be universal across git providers used in conjunction with Flux. A new field, `Git`, will be added, and validations will ensure that `Git` and `Github` (and any other Provider type) are mutually exclusive.
+```Go
+type Flux struct {
+    // Github is the name of the Git Provider to host the Git repo.
+	Github Github `json:"github,omitempty"`
+
+    // Git is the name of the generic Git Provider to host the Git repository
+    Git Git `json:"git,omitempty"`
+
+    // FluxSystemNamespace scope for this operation. Defaults to flux-system.
+    FluxSystemNamespace string `json:"fluxSystemNamespace,omitempty"`
+    
+    // ClusterConfigPath relative to the repository root, when specified the cluster sync will be scoped to this path.
+    ClusterConfigPath string `json:"clusterConfigPath,omitempty"`
+
+	// Git branch. Defaults to main.
+	// +kubebuilder:default:="main"
+	Branch string `json:"branch,omitempty"`
+}
+```
+
+The `Git` struct contains the requisite fields for authenticating to a generic git repository using basic auth, either password, SSH or both together. Use of `net/url` for the repository URL provides out-of-the-box validations and assurance of structure.
+```Go
+type Git struct {
+    // Username is the user to authenticate to the git repository with.
+    Username string `json:username`
+
+    // Repository URL for the repository to be used with flux. Can be either an SSH or HTTPS url.
+    RepositoryUrl net/url `json:repositoryUrl`
+
+    // Path to a private key file for authenticating to the repository via SSH.
+    PrivateKeyFile string `json:privateKeyFile,omitempty`
+
+    // An environment variable which should contain a user-provided password for authenticating to the git repository. 
+    passwordEnvironmentVariable string `json:passwordEnvironmentVariable,omitempty`
+}
+```
+
+The `Github` struct will now no longer contain the `ClusterConfigPath` and `FluxSystemNamespace` fields, which are not specific to Github. It will contain only the fields required to establish a connection to Github -- Owner, Repository, and wether or not it's a Personal repository.
```

Comment:
We can deprecate them but not remove

Reply:
Is there a document for release and deprecation process and/or support model we follow?